backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java

```diff
@@ -99,45 +109,86 @@ public LiveOktaRepository(
+    groupNamesToAdd.stream()
+        .filter(n -> !orgGroupNames.contains(n))
+        .forEach(
+            n -> {
+              throw new IllegalGraphqlArgumentException(
+                  String.format("Cannot add Okta user to nonexistent group=%s", n));
+            });
+    Set<String> groupIdsToAdd =
+        orgGroups.stream()
+            .filter(g -> groupNamesToAdd.contains(g.getProfile().getName()))
+            .map(g -> g.getId())
+            .collect(Collectors.toSet());
 
     UserBuilder.instance()
         .setProfileProperties(userProfileMap)
-        .setGroups(groupIds)
+        .setGroups(groupIdsToAdd)
         .buildAndCreate(_client);
 
-    return Optional.of(new OrganizationRoleClaims(organizationExternalId, roles));
+    List<OrganizationRoleClaims> claims = _extractor.convertClaims(groupNamesToAdd);
+    if (claims.size() != 1) {
+      LOG.warn("User is in {} Okta organizations, not 1", claims.size());
+      return Optional.empty();
+    }
+    return Optional.of(claims.get(0));
   }
 
-  public Set<String> getAllUsernamesForOrganization(Organization org, OrganizationRole role) {
-    String groupName = generateGroupName(org.getExternalId(), role);
+  public Map<String, OrganizationRoleClaims> getAllUsersForOrganization(Organization org) {
 
-    GroupList groups = _client.listGroups(groupName, null, null);
-    if (groups.stream().count() == 0) {
-      LOG.warn(
-          "Okta group for org={}, role={} is nonexistent; returning zero usernames",
-          org.getExternalId(),
-          role.name());
-      return Set.of();
+    GroupList orgGroups =
+        _client.listGroups(generateGroupOrgPrefix(org.getExternalId()), null, null);
+    if (orgGroups.stream().count() == 0) {
+      throw new IllegalGraphqlArgumentException(
+          "Cannot get Okta users for org=" + org.getExternalId() + ": Okta groups are nonexistent");
     }
-    Group group = groups.single();
-    return group.listUsers().stream()
-        .map(u -> u.getProfile().getEmail())
-        .collect(Collectors.toSet());
+
+    // Create a map from each user to their Okta groups...
+    Map<String, Set<String>> usersToGroupNames = new HashMap<>();
+    orgGroups.forEach(
+        g -> {
+          String groupName = g.getProfile().getName();
+          g.listUsers()
+              .forEach(
+                  u -> {
+                    String username = u.getProfile().getEmail();
+                    if (!usersToGroupNames.containsKey(username)) {
+                      usersToGroupNames.put(username, new HashSet<>());
+                    }
+                    usersToGroupNames.get(username).add(groupName);
+                  });
+        });
+
+    // ...then convert each user's relevant Okta groups to an OrganizationRoleClaims object
+    return usersToGroupNames.entrySet().stream()
+        .collect(
+            Collectors.toMap(e -> e.getKey(), e -> _extractor.convertClaims(e.getValue()).get(0)));
```

Comment:
I think it's probably worth adding a method to the extractor that does the "there must be exactly one" check/select internally, if we're going to use it in this context (which I support!). The get-the-first-item-of-the-returned-list pattern is nerve-wracking to look at even when I know it's safe by construction.